gpflow/base.py

```diff
@@ -1,51 +1,41 @@
 import functools
 from enum import Enum
-from typing import List, Optional, Tuple, Union
+from typing import Any, List, Optional, Sequence, Tuple, TypeVar, Union
 
 import numpy as np
 import tensorflow as tf
 import tensorflow_probability as tfp
+from IPython.lib import pretty
 from tensorflow.python.ops import array_ops
 
 from .config import default_float
 
 DType = Union[np.dtype, tf.DType]
-VariableData = Union[List, Tuple, np.ndarray, int, float]
-TensorLike = (
-    object  # Union[tf.Tensor, tf.Variable, np.ndarray], but doesn't work with multipledispatch
-)
+VariableData = Union[List, Tuple, np.ndarray, int, float]  # deprecated
+_TensorLikeExternalTypes = Union[tf.Tensor, tf.Variable, np.ndarray]  # todo waiting on TensorTypes from other PR for multipledispatch to work
+
 Transform = Union[tfp.bijectors.Bijector]
 Prior = Union[tfp.distributions.Distribution]
 
+S = TypeVar("S")
+_NestedSeq = Union[
+    Sequence[S],
+    Sequence[Sequence[S]],
+    Sequence[Sequence[Sequence[S]]],
+    Sequence[Sequence[Sequence[Sequence[Any]]]]  # catch all for rank > 3
+]
+_NativeScalar = Union[int, float]
+_ArrayOrScalar = Union[_NativeScalar, _NestedSeq[_NativeScalar]]
```

Comment:
This is a case where I find the added typing significantly reduces code readability.

Reply:
I can add a docstring for `_ArrayOrScalar` explaining what it is

Reply:
added comment. how's that?